### PR TITLE
New Build Setup

### LINKED ISSUE #321: Zombie serve process
Running into port 3000 conflicts occasionally, only to discover a background serve process that wasn't killed properly. Need to track down and kill all zombies.

### LINKED ISSUE #325: App serve should only warn and then continue for most lint issues
Currently you have to fix all lint issues before the app serve will show the page, which is pretty annoying.

The cause is a clash between `react-scripts` and spotify `web-scripts` where the former uses warning to signal lint issues that won't cause issues during development, but `web-scripts` lists all rules as errors.

I'm thinking we might be able to do more merging of the lint rules, or ask `web-scripts` to adopt a similar pattern.

### LINKED ISSUE #658: Switch app:serve and app:build to webpack build instead of react-scripts
We're currently using `react-scripts` for the app build, but it doesn't have quite the config we want.

We should switch to a custom webpack config, shipped as a part of the CLI. It should be shared with https://github.com/spotify/backstage/blob/master/packages/cli/src/commands/plugin/serve/config.ts so that plugin serve and app serve supports the same leaders etc.

The difference to the existing plugin:serve config is that we need to support a production build for the app as well. The production build should be optimized for the plugin structure in backstage, lifting out what we're currently using in the internal version.

### LINKED ISSUE #669: [CLI] Keep JSDoc comments in type definition files
## Feature Suggestion

We should keep JSDoc comments intact in `.d.ts` files in `dist/`. Right now they're being stripped out because ”removeComments" is set to true in the tsconfig inherited from `@spotify/web-scripts`

## Possible Implementation

We likely have to switch off "removeComments" in the cli/config, or override it in the rollup plugin. Although since we don'y want the JS files to have a bunch of comments, we probably have to add a rollup option or plugin to remove comments from the js file.

## Context

We want doc popups in our editors!

### LINKED ISSUE #834: [RFC] New Build Setup
## Context

The current build setup ranges from pretty slow, for example bundling, to extremely slow, looking at ~1min core package builds. There's plenty of space for improvement, but it's been unclear exactly what build setups would be best, especially given that we want a setup that scales to much larger projects than this is right now.

To figure out the options for different parts of the builds setups and what tradeoffs we'd actually been making, I've been doing a bunch of benchmarks in this project: [Rugvip/ts-build-bench](https://github.com/Rugvip/ts-build-bench).

## Benchmarking Outcome

The benchmarking was divided into 3 major categories, VSCode performance, bundle/build speed, and type checking speed. For a more in-depth explanation, see the [ts-build-bench README](https://github.com/Rugvip/ts-build-bench/blob/master/README.md).

The outcome of the VSCode benchmarking was that there's not much to do, just point straight to src files for development.

The bundling benchmarks showed that even for a very large project, new transpilers like [esbuild](https://github.com/evanw/esbuild) and [sucrase](https://github.com/alangpierce/sucrase) provide enough performance that it's possible to transpile an entire project from TypeScript source files in one go, and still provide a good development experience. Transpilers like `ts-loader` and `babel-loader` are not really quick enough for a good development experience, babel being the one we use in our large internal project right now. If we have to use either of those it'd likely be best to keep the separate builds of each package that we have right now.

For building individual packages with rollup, esbuild and sucrase both provide a huge performance boost over what we're using right now. I also found that esbuild also produces code that is much quicker for webpack to process compared to sucrase.

The outcome of the type checking was that [TypeScript project references](https://www.typescriptlang.org/docs/handbook/project-references.html) can provide a big speedup for setups that require separate TypeScript configurations for each package. But, by far the quickest method, and the only reasonable one for a huge project, is to use a top-level tsconfig that lints the entire project, basically pointing to every `src/` directory of every package. The downside of this top-level approach is that declaration files end up in a separate directory and need to be moved to their respective packages before publishing.

## New Suggested Setup

Based on the outcome of the benchmarking and reading up on various tools, I'm suggesting the following changes:

1. Point all `types` fields in `package.json` to `src/index.ts`, published packages still point to `dist/index.d.ts`.
1. Switch `plugin:build` to use rollup with esbuild plugin, but only use build for publishing packages. Minify output as well, if it works. No need to support watch mode either.
1. Switch `plugin:serve` and `app:serve` to both use a common custom webpack config. Use sucrase-loader as it has react-hot-loader support, and have esbuild-loader as a fallback in case sucrase-loader doesn't work out. Use the same loader for both development and production.
1. Skip type-checking in all build steps, and rely on editors and `tsc` instead. Start using a single top-level tsconfig for linting, use incremental mode and an `outDir` that is cached in CI. Use [rollup-plugin-dts](https://www.npmjs.com/package/rollup-plugin-dts) or whatever other `.d.ts` bundling tool is good to build a single `.d.ts` file for each package. If they don't work well, just copy `.d.ts` files from the common output dir to each package before publishing.

## Reasoning

Some thoughs and constraints explaining the above choices:

1. We don't accept an average of 1-2 minutes of waiting to start up a dev server, if it's that slow we need to find other solutions.
1. Production builds should be relatively quick, no more than a few minutes including minification.
1. The build setup should support hundreds of plugins. This project will grow much larger, and we want to use the same setup for our internal version of Backstage as well.
1. We want to keep webpack because it has excellent bundle splitting strategies that keeps the initial page load very quick.
1. It is better to use the same transilers for development and production builds. I don't think we want to use for example sucrase for development but babel for production builds.
1. We're targetting modern and up-to-date browsers, so transpilation is mostly a matter of removing types and transforming JSX.
1. Users of backstage that want to target older browsers can set up a custom webpack build for that purpose.
1. The project is still in alpha so we can choose a build strategy and if it doesn't work out, switch.

### PR DESCRIPTION
**Pull request by Rugvip**
Implements the rest of #834, couldn't really find a nice way to split this apart, sorry about that :grin:

Main things that have changed:

* Everything is much faster 🏎️💨
* React hot loader 🔥 
* No building dependencies during development, everything is read from `src/`, this includes types, dev server, jest tests, storybook. 🛸 
* Type checking is done at the top-level of the project only, with incremental mode enabled. This means packages can't have their own TS config, but it also means type checking is and order of magnitude faster. ✔️ 
* Type declaration files are built into a single file per package, with comments. 📦 
* We no longer need to depend on react-scripts 🎉 
* Type checking and linting during builds is optional, rely on the editor instead. 📝 

Not changing the backend build yet, as that was something that could be kept out of scope :grin:

There's an ugly hack with patching node_modules to get types to work, hopefully we can get rid of that asap. There's a readme in the folder which explains it.

Fixes #834, fixes #325, fixes #658, fixes #321, fixes #669

### CODE REVIEW

File: packages/core/src/index.ts [hunk 1]
```diff
@@ -17,7 +17,8 @@
 export * from './api';
 export { default as Page } from './layout/Page';
 export { gradients, pageTheme } from './layout/Page';
-export type { PageTheme } from './layout/Page';
+import type { PageTheme } from './layout/Page';
+export type { PageTheme };
```
Comment (Rugvip) [inline packages/core/src/index.ts:21]: We can revert all of these once https://github.com/alangpierce/sucrase/pull/533 is released

Comment (Rugvip) [reply packages/core/src/index.ts]: ... which it has been now, so changes reverted :grin:

File: packages/cli/src/lib/tasks.ts [hunk 2]
```diff
@@ -103,3 +107,92 @@ export async function templatingTask(
...
+        'patching',
+        `node_modules/@backstage/${name} package.json`,
+        async () => {
+          const depJsonPath = resolvePath(
+            dir,
+            'node_modules/@backstage',
+            name,
+            'package.json',
+          );
+          console.log('DEBUG: depJsonPath =', depJsonPath);
```
Comment (Rugvip) [inline packages/cli/src/lib/tasks.ts:180]:
```suggestion
```

File: packages/cli/src/lib/packager/packager.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import chalk from 'chalk';
+import { relative as relativePath } from 'path';
+import { paths } from '../paths';
+import { makeConfigs } from './config';
+
+function formatErrorMessage(error: any) {
+  let msg = '';
+
+  if (error.code === 'PLUGIN_ERROR') {
+    // typescript2 plugin has a complete message with all codeframes
```
Comment (Rugvip) [inline packages/cli/src/lib/packager/packager.ts:27]:
```suggestion
```

**Review by marcuseide — APPROVED** [review]

**Pull request merged**